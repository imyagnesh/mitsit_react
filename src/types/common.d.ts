type LoadingAppStatus = {
  state: "LOADING";
  error?: never;
};

type SuccessAppStatus = {
  state: "SUCCESS";
  error?: never;
};

type ErrorAppStatus = {
  state: "ERROR";
  error: Error;
};

type WithoutIdType = "LOAD_TODO" | "ADD_TODO";

type WithIdType = "UPDATE_TODO" | "DELETE_TODO";

type WithoutIdTypes = {
  type: WithoutIdType;
  id?: never;
};

type WithIdTypes = {
  type: WithIdType;
  id: number;
};

type BaseAppStatus = LoadingAppStatus | ErrorAppStatus | SuccessAppStatus;

type BaseTypeStatus = WithoutIdTypes | WithIdTypes;

export type AppStatus = BaseAppStatus & BaseTypeStatus;

export type StatusType = WithoutIdType | WithIdType;
