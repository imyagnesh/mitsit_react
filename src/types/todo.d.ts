export type TodoItemType = {
  id: number;
  text: string;
  isDone: boolean;
};

export type FilterType = "all" | "pending" | "completed";
