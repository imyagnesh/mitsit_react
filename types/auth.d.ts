declare type AuthResponse = {
  accessToken: string;
  user: User;
};
