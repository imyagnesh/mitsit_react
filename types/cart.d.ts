declare type CartType = {
  id: number;
  quantity: number;
  productId: number;
};
