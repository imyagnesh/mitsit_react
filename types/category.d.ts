declare enum CategoryEnum {
  MensClothing = `men's clothing`,
  Jewelry = 'jewelry',
  Electronics = 'electronics',
  WomansClothing = `women's clothing`,
}
