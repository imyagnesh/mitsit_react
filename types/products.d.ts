declare type ProductType = {
  id: number;
  title: string;
  price: number;
  description: string;
  category: CategoryEnum;
  image: string;
  rating: RatingType;
};
