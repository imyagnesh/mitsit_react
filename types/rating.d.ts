declare type RatingType = {
  rate: number;
  count: number;
};
