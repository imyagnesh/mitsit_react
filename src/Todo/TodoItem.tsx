import React, { memo } from "react";
import cn from "classnames";
import { AppStatus, TodoItemType } from ".";

type Props = {
  todoItem: TodoItemType;
  completeTodo: (todoItem: TodoItemType) => void;
  deleteTodo: (id: number) => void;
  status: AppStatus | undefined;
};

const TodoItem = ({ todoItem, completeTodo, deleteTodo, status }: Props) => {
  // console.log("Todo Item render");

  return (
    <div key={todoItem.id} className="flex items-center m-2">
      <input
        type="checkbox"
        disabled={status?.type === "UPDATE_TODO" && status.state === "LOADING"}
        checked={todoItem.isDone}
        onChange={() => completeTodo(todoItem)}
        className="focus:ring-indigo-500 h-4 w-4 text-indigo-600 border-gray-300 rounded disabled:text-gray-600 disabled:ring-gray-500"
      />
      <p
        className={cn("flex-1 px-4", {
          "line-through": todoItem.isDone,
        })}
      >
        {todoItem.text}
      </p>
      <button
        disabled={status?.type === "DELETE_TODO" && status.state === "LOADING"}
        className="btn disabled:bg-slate-400"
        onClick={() => deleteTodo(todoItem.id)}
      >
        Delete
      </button>
    </div>
  );
};

export default memo(TodoItem);
