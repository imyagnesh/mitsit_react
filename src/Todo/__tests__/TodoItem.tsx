import * as React from "react";
import {
  fireEvent,
  render as renderRtl,
  RenderResult,
  screen,
} from "@testing-library/react";
import TodoItem from "../TodoItem";
import { TodoItemType } from "../../types/todo";

const todoItem: TodoItemType = {
  text: "get milk",
  isDone: true,
  id: 1,
};

const completeTodo = jest.fn();
const deleteTodo = jest.fn();

const render = ({ ...props } = {}) => {
  const ui = (
    <TodoItem
      todoItem={todoItem}
      completeTodo={completeTodo}
      deleteTodo={deleteTodo}
      {...props}
    />
  );
  return renderRtl(ui);
};

describe("TodoItem Component", () => {
  let renderItem: typeof render;

  test("should render TodoItem", () => {
    // const container = screen.queryByTestId("todo_item_container");
    const { container } = render();

    expect(container.firstChild).toBeInTheDocument();
  });

  test("should all the elements are available", () => {
    render();
    const deleteBtn = screen.queryByRole("button", { name: "Delete" });
    expect(deleteBtn).toBeInTheDocument();
    const check = screen.queryByRole("checkbox", {
      checked: todoItem.isDone,
    });
    expect(check).toBeInTheDocument();
    expect(deleteTodo).toBeCalledTimes(0);
    const text = screen.queryByText(todoItem.text);
    expect(text).toBeInTheDocument();
  });

  test("should click delete button", () => {
    const { debug } = render();
    const deleteBtn = screen.queryByRole("button", { name: "Delete" });
    if (deleteBtn) {
      fireEvent.click(deleteBtn);
      expect(deleteTodo).toBeCalledTimes(1);
      expect(deleteTodo).toBeCalledWith(todoItem);
    }
  });

  test("should complete todo", () => {
    const { debug, rerender } = render();
    const checkbox = screen.queryByRole("checkbox", { checked: true });

    if (checkbox) {
      fireEvent.click(checkbox);
      expect(completeTodo).toBeCalledTimes(1);
      expect(completeTodo).toBeCalledWith(todoItem);
      rerender(
        <TodoItem
          todoItem={{ ...todoItem, isDone: false }}
          completeTodo={completeTodo}
          deleteTodo={deleteTodo}
        />
      );
      expect(checkbox).not.toBeChecked();
    }
  });

  test("should disable delete button", () => {
    render({
      status: {
        type: "DELETE_TODO",
        state: "LOADING",
      },
    });

    const deleteBtn = screen.queryByRole("button", { name: "Delete" });
    expect(deleteBtn).toBeDisabled();
  });

  test("should checkbox disabled", () => {
    render({
      status: {
        type: "UPDATE_TODO",
        state: "LOADING",
      },
    });

    const checkbox = screen.queryByRole("checkbox", { checked: true });
    expect(checkbox).toBeDisabled();
  });
});
