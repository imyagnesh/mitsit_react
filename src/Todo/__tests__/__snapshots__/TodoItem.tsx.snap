// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem Component should compare snapshot 1`] = `
<div
  class="flex items-center m-3"
  data-testid="todo_item_container"
>
  <input
    checked=""
    class="focus:ring-indigo-500 h-4 w-4 text-indigo-600 border-gray-300 rounded disabled:text-gray-600 disabled:ring-gray-500"
    type="checkbox"
  />
  <p
    class="flex-1 px-4 line-through"
  >
    get milk
  </p>
  <button
    class="btn disabled:bg-slate-400"
  >
    Delete
  </button>
</div>
`;
